src/instruction.rs

```diff
@@ -0,0 +1,268 @@
+//! Program instruction types.
+
+use solana_program::{
+    instruction::{AccountMeta, Instruction},
+    program_error::ProgramError,
+    pubkey::Pubkey,
+};
+
+/// Instructions supported by the Paladin Rewards program.
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum PaladinRewardsInstruction {
+    /// Configures rewards for a mint that has been configured with the rewards
+    /// program as a transfer hook program.
+    ///
+    /// This instruction will:
+    ///
+    /// - Initialize a rewards account for the mint (distribution account),
+    ///   configured with the distribution addresses:
+    ///   - Active rewards
+    ///   - Piggy bank
+    ///   - Staked PAL rewards
+    /// - Initialize an active rewards account for the mint.
+    /// - Initialize the required accounts for the transfer hook.
+    ///
+    /// Accounts expected by this instruction:
+    ///
+    /// 0. `[w]` Distribution account.
+    /// 1. `[w]` Active rewards account.
+    /// 2. `[w]` Transfer hook extra account metas account.
+    /// 3. `[ ]` Token mint.
+    /// 4. `[s]` Mint authority.
+    InitializeMintRewardInfo {
+        piggy_bank_address: Pubkey,
+        staked_rewards_address: Pubkey,
+    },
+    /// Moves the active rewards to the distribution account and updates total
+    /// rewards.
+    ///
+    /// Accounts expected by this instruction:
+    ///
+    /// 0. `[w]` Active rewards account.
+    /// 1. `[w]` Distribution account.
+    /// 2. `[ ]` Token mint.
+    UpdateMintRewards,
+    /// Moves SOL rewards to the following parties:
+    ///
+    /// - 1%  Piggy bank
+    /// - 4%  Staked PAL (validators)
+    /// - 5%  All PAL holders
+    /// - 90% Leader who produces the block
+    ///
+    /// Accounts expected by this instruction:
+    ///
+    /// 0. `[w]` Active rewards account.
+    /// 1. `[w]` Distribution account.
+    /// 2. `[w]` Piggy bank account.
+    /// 3. `[w]` Staked PAL rewards account.
+    /// 4. `[w]` Leader account.
+    DistributeRewards,
```

Comment:
how do we plan to store the distribution configuration? (in case it were to change in the future via governance or similar)

Reply:
We could introduce a new PDA for this configuration, but is there some keyholder you'd want to hold the update authority? And where do we store that pubkey, hard-coded?

Reply:
@edgarpavlovsky Looping back to this idea, in case you want to introduce some additional state.